modules/data.land/R/soil_params_ensemble.R

```diff
@@ -0,0 +1,224 @@
+  estimate_alpha0 <- function(means, quantiles) {
+    # Objective function to minimize the difference between observed and simulated quantiles with means as a known moment
+    objective_function <- function(alpha0) {
+      if (alpha0 <= 0)
+        return(Inf)
+      alpha <- means * alpha0
+      q5 <- quantiles$q5
+      q50 <- quantiles$q50
+      q95 <- quantiles$q95
+      
+      # Compute differences with observed quantiles
+      samples <- MCMCpack::rdirichlet(10000, alpha) # Generate samples
+      sim_q5 <- apply(samples, 2, quantile, probs = 0.05, na.rm = TRUE)
+      sim_q50 <-apply(samples, 2, quantile, probs = 0.50, na.rm = TRUE)
+      sim_q95 <-apply(samples, 2, quantile, probs = 0.95, na.rm = TRUE)
+      quantile_diff <-sum((sim_q5 - q5) ^ 2 + (sim_q50 - q50) ^ 2 + (sim_q95 - q95) ^ 2)
+      return(quantile_diff)
+    }
+    
+    # Optimize alpha0
+    result <- optim(
+      par = 1, # Initial guess for alpha0
+      fn = objective_function,
+      method = "L-BFGS-B",
+      lower = 0.01  # alpha0 must be positive
+    )
+    return(result$par)
+  }
+ 
+  alpha0 <- estimate_alpha0(means, quantiles)
+  if (alpha0 <= 0) {
+    stop("Estimated alpha0 is non-positive, which is invalid.")
+  }
+  alphas <- means * alpha0
+  return(alphas)
+}
+
+
+
+##' A function to estimate the soil parameters based on soil texture data and write the parameter paths into settings
+##' 
+##' @param settings A multi-site settings
+##' @param sand A data frame containing sand fraction in percentage from SoilGrids250m v2.0 with columns "Depth", "Quantile", "Siteid", and "Value"
+##' @param clay A data frame containing clay fraction in percentage from SoilGrids250m v2.0 with columns "Depth", "Quantile", "Siteid", and "Value"
+##' @param silt A data frame containing silt fraction in percentage from SoilGrids250m v2.0 with columns "Depth", "Quantile", "Siteid", and "Value"
+##' @param outdir Provide the path to store the parameter files
+##' @param write_into_settings Whether to write the path of parameter file into the setting. The default is TRUE
+##' 
+##' @example
+##' \dontrun{
+##' 
+##' outdir <- "/projectnb/dietzelab/Cherry/SoilGrids_texture/39NEON"
+##' sand <- readRDS("/projectnb/dietzelab/Cherry/SoilGrids_texture/sand_percent.rds") #sand fraction in percentage
+##' clay <- readRDS("/projectnb/dietzelab/Cherry/SoilGrids_texture/clay_percent.rds") #clay fraction in percentage
+##' silt <- readRDS("/projectnb/dietzelab/Cherry/SoilGrids_texture/silt_percent.rds") #silt fraction in percentage
+##' settings <-read.settings("/projectnb/dietzelab/Cherry/xml/pecan_monthly_SDA_soilwater.xml")
+##' soil_params_ensemble(settings,sand,clay,silt,outdir)
+##'  }
+##'  
+##'  @return Ensemble soil parameter files defined in outdir and file paths in xml file
+##'  @export 
+##'  @author Qianyu Li
+##'  @importFrom magrittr %>%
+##'  
+
+soil_params_ensemble <- function(settings,sand,clay,silt,outdir,write_into_settings=TRUE){
```

Comment:
Conceptually, it would be nice to have a more clean separation between things that are specific to one product vs general. Right now this function is specific to SoilGrids. I think you might want to either generalize the function a bit more (e.g. so that it would still work with a different soil map with different layer thicknesses and different data structures) or modify the function name to make it clear that it is only for this one product.

Reply:
Fixed with change of function name.